torch/csrc/jit/passes/memory_planning/memory_observer.cpp

```diff
@@ -0,0 +1,165 @@
+//
+// Created by Maksim Levental on 9/11/21.
+//
+
+#include <iostream>
+#include <utility>
+
+#include <torch/csrc/jit/ir/ir.h>
+#include <torch/csrc/jit/passes/memory_planning/memory_observer.h>
+#include <torch/csrc/jit/runtime/interpreter.h>
+
+namespace torch {
+namespace jit {
+
+MemoryObserverThreadLocalState* getMemoryObserverTLSState() {
+  const auto& state =
+      c10::ThreadLocalDebugInfo::get(c10::DebugInfoKind::PROFILER_STATE);
+  return static_cast<MemoryObserverThreadLocalState*>(state);
+}
+
+struct MemoryObserverContext : public at::ObserverContext {
+  MemoryObserverContext(
+      std::vector<intptr_t> input_ival_addrs,
+      std::vector<std::string> input_val_names,
+      std::vector<std::string> output_val_names,
+      std::string fn_name,
+      int64_t start_time)
+      : input_ival_addrs(std::move(input_ival_addrs)),
+        input_val_names(std::move(input_val_names)),
+        output_val_names(std::move(output_val_names)),
+        fn_name(std::move(fn_name)),
+        start_time(start_time) {}
+  std::vector<intptr_t> input_ival_addrs;
+  std::vector<std::string> input_val_names;
+  std::vector<std::string> output_val_names;
+  std::string fn_name;
+  int64_t start_time;
+};
+
+void pushObserverCallbacks() {
+  auto state_ptr = getMemoryObserverTLSState();
+  TORCH_INTERNAL_ASSERT(state_ptr, "Expected memory observer state set");
+  auto start_cb =
+      [](const at::RecordFunction& fn) -> std::unique_ptr<at::ObserverContext> {
+    auto state_ptr = getMemoryObserverTLSState();
+    if (!state_ptr) {
+      return nullptr;
+    }
+    std::lock_guard<std::mutex> guard(state_ptr->state_mutex);
+
+    std::vector<intptr_t> input_ival_addrs;
+    for (size_t i = fn.inputs().size() - fn.num_inputs();
+         i < fn.inputs().size();
+         i++) {
+      auto ival = fn.inputs()[i];
+      input_ival_addrs.emplace_back(
+          ival.isTensor()
+              ? reinterpret_cast<intptr_t>(ival.toTensor().data_ptr())
+              : 0);
+    }
+
+    std::vector<std::string> input_val_names;
+    std::vector<std::string> output_val_names;
+    if (jit::currentFrameId()) {
```

Comment:
when would we not expect this to be set?

Reply:
not only are there instances where it won't be set but there are times when constructing it is ill-posed. I didn't track it down but it's roughly at times before the interpreter has built a complete frame (I was hitting segfaults before i girded these loins)